crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2603,6 +2604,59 @@ impl<'db> TypeInferenceBuilder<'db> {
                     op,
                     Type::IntLiteral(i64::from(b)),
                 ),
+
+            (Type::StringLiteral(salsa_s1), Type::StringLiteral(salsa_s2)) => {
+                // For anything non-ASCII, don't even try to infer precise literal types.
+                // It would be complicated, it's an unimportant edge case,
+                // and Rust and Python probably do subtly different things
+                // in their string methods for non-ASCII characters.
+                let s1 = salsa_s1.value(self.db);
+                if !s1.is_ascii() {
```

Comment:
Sorry to get back to this only after the change was made, but I feel pretty strongly we should not do this. It's discriminatory against people using non ASCII strings 😆 seriously, it will be surprising if your type inference changes just because you add a non ASCII character in a string literal. We should provide the same level of behavior for all string literals that we support at all. If there's an operation we don't believe we can reasonably support for all string literals, we should just support it for none of them. 

At the very least I think it's important that we do support equality and inequality here. Salsa already interns our string types based on rust string equality, so if it differs from Python string equality we have an issue that goes beyond this code. 

I feel less strongly about comparison. But my preference would be to just support them all, assuming rust/python equivalence, and leave a note or TODO comment to dig deeper for possible edge cases.

Reply:
Fair enough, I withdraw this suggestion! I, on the other hand, do feel pretty strongly that we should actually check for corner cases here where Rust and Python differ in their handling of strings ;) So I would like us to at least leave a TODO comment (rather than just a note) to check this properly. Otherwise, as I say, I feel like it's inevitable that we'll get bug reports around this. Obviously it doesn't come up that much in real-world code, but people love to hunt for this kind of corner case in static-analysis tools and report them 😆 And, you know, accuracy is nice!

Reply:
I'm completely for a `// TODO`, won't look into this myself just yet though

Reply:
I did a bit of looking, and it seems like both Python and rust have a pretty simple algorithm for the default string comparison: it's just a simple lexicographic sort by Unicode code points. (The complexities happen when you get into locale aware sorting, which requires special libraries in either language.) And we should have the same Unicode code points in our literal strings, since we're literally reading the same source text. (If we don't, that's a bug elsewhere, not here.)

So my conclusion is that we can just go ahead and do this, and there's no need for either a note or a todo comment! Sorry for even raising the question unnecessarily instead of doing the research the first time 😆

Reply:
We could have done the research as well, so the blame is shared :)